File: LeetCode/708.Insert_into_a_Sorted_Circular_Linked_List/708.Insert_into_a_Sorted_Circular_Linked_List.py
#!/usr/bin/env python3

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    def insert(self, head: "Optional[Node]", insertVal: int) -> "Node":
        # non-decreasing order
        # After the insertion, the circular list should remain sorted.

        # head is null, create a new single circular list
        insert = Node(insertVal)
        if not head:
            insert.next = insert
            return insert

        cur = head
        while True:
            # 1 [2] 3
            if cur.val <= insertVal <= cur.next.val:
                break
            # [1] 2 3 or 2 3 [4] cur is biggest, cur.next is smallest
            if cur.val > cur.next.val:
                if cur.val < insertVal or insertVal < cur.next.val:
                    break

            cur = cur.next
            # finished one pass
            if cur == head:
                break

        insert.next = cur.next
        cur.next = insert

        return head

File: LeetCode/708.Insert_into_a_Sorted_Circular_Linked_List/708.Insert_into_a_Sorted_Circular_Linked_List.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var next: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func insert(_ head: Node?, _ insertVal: Int) -> Node? {
        // handle head is nil case
        let insertNode = Node(insertVal)
        guard let head = head else {
            insertNode.next = insertNode
            return insertNode
        }

        // only 1 node
        if head.next == head {
            head.next = insertNode
            insertNode.next = head
            return head
        }

        // two pointer, one is prev, one is current
        // prev.val <= insertVal < curr.val

        // prev.val > cur.val
        var prev: Node? = head.next, curr: Node? = head
        var insert = false

        while let p = prev, let c = curr {
            // prev.val > insertVal >= curr.val
            if c.val <= insertVal, insertVal < p.val {
                c.next = insertNode
                insertNode.next = p
                insert = true
                break
            }

            // insertVal >= all vals in linked list / <= all vals in linked list
            if p.val < c.val, insertVal <= p.val || insertVal >= c.val {
                c.next = insertNode
                insertNode.next = p
                insert = true
                break
            }

            prev = p.next
            curr = c.next

            if curr == head { break }
        }

        if !insert {
            curr?.next = insertNode
            insertNode.next = prev
        }

        return head
    }
}
